File: go/boardgame/boardgame.go
package boardgame

import "fmt"

type Game interface {
    PlayAt(x, y int)
    FlagAt(x, y int)
    ValueAt(x, y int) string
    Cols() int
    Rows() int
    Finished() bool
    Message() string
}

func ConsolePrint(game Game) {
    for y := 0; y < game.Rows(); y++ {
        for x := 0; x < game.Cols(); x++ {
            fmt.Printf("%3s", game.ValueAt(x, y))
        }
        fmt.Println()
    }
}

func ConsolePlay(game Game) {
    ConsolePrint(game)

    for ! game.Finished() {
        var x, y int
        fmt.Scan(&x, &y)
        game.PlayAt(x, y)
        ConsolePrint(game)
    }

    fmt.Println(game.Message())
}

File: fifteen/boardgame.py
#!/usr/bin/env python3
'''
@author  Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

def abstract():
    raise NotImplementedError("Abstract method")

class BoardGame:
    def play_at(self, x: int, y: int): abstract()
    def flag_at(self, x: int, y: int): abstract()
    def value_at(self, x: int, y: int) -> str: abstract()
    def cols(self) -> int: abstract()
    def rows(self) -> int: abstract()
    def finished(self) -> bool: abstract()
    def message(self) -> str: abstract()


def print_game(game: BoardGame):
    for y in range(game.rows()):
        for x in range(game.cols()):
            print(game.value_at(x, y), end="\t")
        print()

def console_play(game: BoardGame):
    print_game(game)

    while not game.finished():
        x, y = input().split()
        game.play_at(int(x), int(y))
        print_game(game)

    print(game.message())
